pkg/scheduler/core/scheduling_queue.go

```diff
@@ -312,6 +330,12 @@ func (p *PriorityQueue) Pop() (*v1.Pod, error) {
 	p.lock.Lock()
 	defer p.lock.Unlock()
 	for len(p.activeQ.data.queue) == 0 {
+		// When the queue is empty, invocation of Pop() is blocked until new item is enqueued.
+		// When Close() is called, the p.closed is set and the condition is broadcast,
+		// which causes this loop to continue and return from the Pop().
+		if p.closed {
+			return nil, fmt.Errorf(queueClosed)
+		}
 		p.cond.Wait()
 	}
 	obj, err := p.activeQ.Pop()
```

Comment:
Should also check queue is closed.

Reply:
Hmm, I doubt it. `p.activeQ.Pop()` should be consistent with `len(p.activeQ.data.queue) == 0`, right? As we have mutex protecting this.

Reply:
when `p.cond.Wait()` return , we need to recheck the status queue as this thread may not the first one to get the lock; so how about some other thread add more item before this thread get the lock?

Reply:
This is a good point.

I read some article saying that when `cond.Wait()` is called, the goroutine is suspended, but it's smart to not hold the lock. Instead, it tries to re-acquire the lock when it's awaken by a `Broadcast()` or `Signal()`.

The article is [here](https://kaviraj.me/understanding-condition-variable-in-go/), and it states like this:

> It may look like that the goroutine waiting(rec.cond.Wait()) is holding the lock whole time(rec.Lock()), but its not, Internally cond.Wait() unlocks it and it **locks it again** only when it wakes up by other go routine.

So if this's true, when it's awaken again, it should try to acquire the lock first, and then check if its length is 0, then it's safe.

But I need to write some sample to verify above saying.

Reply:
> So if this's true, when it's awaken again, it should try to acquire the lock first

When `wait()` return, it will get the lock again; but we do not know what was happened between `Broadcast`/`Signal` and `getting lock again`. That's why we use `for loop` to check `queue.len` again.

> and then check if its length is 0, then it's safe.

That's what the `for loop` doing :)

Reply:
In this case, if queue is closed but there are still new items (added by other goroutines), then without checking if queue is closed, resting items will still be handled.

So @k82cn your preference is to check if queue is closed, and return immediately if it is? (as it's sort of wasting time handling those items.)

Reply:
> your preference is to check if queue is closed, and return immediately if it is?

maybe a `IsClosed` is better; so the code who uses `SchedulingQueue` can decide whether continue to handle the pending item or ignore them.

Reply:
I rethink it a little bit: given that FIFO implementation (in client-go) "drains" the queue, and currently PriorityQueue behaves the same, I think we'd better keep them as is. Any objection? cc @misterikkit @bsalamat

Reply:
> External users are not able to use SchedulingQueue any more lol... As we recently move it to package `internal`.

Anything in `pkg/scheduler/...` is not considered external from the perspective of `pkg/scheduler/internal/...` 
https://golang.org/doc/go1.5

> currently PriorityQueue behaves the same
I like that the two implementations are consistent, and I'm fine with the existing behavior. Let's just make sure to document it as an official part of the `SchedulingQueue` API so that people in the future know what is intended.

Reply:
Is there a chance this could cause a busy loop during shutdown?

Reply:
Even there is, I think it's the same case as before. It doesn't make things worse.

Reply:
One possible case I can think of is that master scheduler lost its leader role, and want to exit gracefully, and _maybe_ there happens to be a long queue.

Reply:
I'm not too worried about that case. If scheduler is shutting down, then it should exit the run loop after Pop returns. We also don't acquire leadership twice in the scheduler (by design). When we lose leadership, the scheduler exits and is restarted.

Reply:
Thanks for humoring me

Reply:
There is no problem actually. Resting items will be handled even if queue is closed.

Reply:
Sorry, I re-read the code, my previous statement is inaccurate. It should be "Resting items will be handled one time at most if queue is closed."

It's because in Pop(), if there are other items enqueued, next item will be popped out. And in outer logic, wait.Until() will return, gracefully.